lib/workflow/task-graph.js

```diff
@@ -0,0 +1,493 @@
+                var waitOnTaskValue = taskData.waitOn[waitOnTask];
+                delete taskData.waitOn[waitOnTask];
+                taskData.waitOn[newWaitOnTaskKey] = waitOnTaskValue;
+            });
+            var taskOverrides = {
+                instanceId: idMap[taskData.label],
+                waitingOn: taskData.waitOn,
+                ignoreFailure: taskData.ignoreFailure
+            };
+
+            if (taskData.taskName) {
+                return self.constructTaskObject(taskData.taskName, taskOverrides,
+                        taskData.optionOverrides, taskData.label)
+                .then(function(definition) {
+                    self.tasks[definition.instanceId] = definition;
+                });
+            } else if (taskData.taskDefinition) {
+                var definition = self.constructInlineTaskObject(taskData.taskDefinition,
+                    taskOverrides, taskData.optionOverrides, taskData.label);
+                self.tasks[definition.instanceId] = definition;
+            }
+        })
+        .spread(function() {
+            return self;
+        });
+    };
+
+    TaskGraph.prototype.constructInlineTaskObject = function(_definition, taskOverrides,
+           optionOverrides, label) {
+
+        return this._buildTaskDefinition(_definition, optionOverrides,
+                taskOverrides, label);
+    };
+
+    TaskGraph.prototype.constructTaskObject = function(taskName, taskOverrides,
+           optionOverrides, label) {
+        var self = this;
+        return store.getTaskDefinition(taskName)
+        .then(function(taskDefinition) {
+            return self._buildTaskDefinition(taskDefinition,
+                    optionOverrides, taskOverrides, label);
+        });
+    };
+
+    TaskGraph.prototype._buildTaskDefinition = function(_definition, optionOverrides,
+            taskOverrides, label) {
+        var self = this;
+        var definition = _.cloneDeep(_definition);
+
+        var baseTaskDefinition = self._getBaseTask(_definition);
+        definition.instanceId = taskOverrides.instanceId;
+        definition.properties = _.merge(definition.properties, baseTaskDefinition.properties);
+        definition.runJob = baseTaskDefinition.runJob;
+        definition.options = _.merge(definition.options || {}, optionOverrides || {});
+        definition.name = taskOverrides.name || definition.injectableName;
+        definition.waitingOn = taskOverrides.waitingOn || {};
+        // TODO: Remove ignoreFailure in favor of better graph branching evalution.
+        // NOTE: actually ignoreFailure is still useful for tasks that can
+        // fail but we don't care AND don't have anything queued up to run
+        // on failure (rare case probably, but still worth supporting IMO).
+        definition.ignoreFailure = taskOverrides.ignoreFailure || false;
+
+        var allOptions = _.uniq(
+                _.keys(definition.options).concat(baseTaskDefinition.requiredOptions));
```

Comment:
Similar reasoning to my above comment, though I'll add an emphasis on how the extra indentation here makes it easier to see/read the variable name by itself as one is scanning through the code. Happy to change it though, just adding some background on why my preferences are they way they are.